File: stack-queue/py/Queue.py


class Node:
    def __init__(self,val = None ,next = None):
        self.val = val
        self.next = next


class Queue:
    def __init__(self,val= None):
        self.head = self.tail = Node(val) if val is not None else None
        self.len = 1 if val is not None else 0

    def enqueue(self,val):

        node = Node(val)
        
        if self.head is None:
           self.head = self.tail = node
        else:
           self.tail.next = node
           self.tail = node

        self.len += 1

    def dequeue(self):
     
        if self.head is None:
            print("Queue is empty")
            return
        
        self.head = self.head.next
        self.len -= 1

        if self.len == 0:
            self.tail = None

    def peek(self):
        return self.head.val if self.head.val else None;
  
    def is_empty(self): 
        return self.len == 0;
  
     
     

my_queue = Queue()
my_queue.enqueue(10)
my_queue.enqueue(20)
my_queue.dequeue()
my_queue.dequeue()
my_queue.dequeue()
print("head",my_queue.head)
print("tail",my_queue.tail)
print("len",my_queue.len)

File: stack-queue/js/Queue.js
class Node {
  constructor(value, next = null) {
    this.value = value;
    this.next = next;
  }
}

class Queue {
  constructor(value) {
    this.head = this.tail = value ? new Node(value) : null;
    this.length = value ? 1 : 0;
  }

  enqueue(value) {
    const node = new Node(value);

    if (!this.head) {
      this.head = this.tail = node;
    } else {
      this.tail.next = node;
      this.tail = node;
    }
    this.length++;
  }

  dequeue() {
    if (!this.head) {
      console.log("Queue is empty");
      return;
    }

    this.head = this.head.next;
    this.length--;

    if (this.length === 0) {
      this.tail = null;
    }
  }

  peek() {
    return this.head ? this.head.value : null;
  }

  isEmpty() {
    return this.length === 0;
  }
}

const myQueue = new Queue();
myQueue.enqueue(10);
myQueue.enqueue(20);
myQueue.dequeue();
myQueue.dequeue();
myQueue.dequeue();

console.log(myQueue);
